File: test/examples/assignment_sat/assignment_sat.cc





























#include <stdlib.h>

#include <vector>

#include "ortools/base/logging.h"
#include "ortools/sat/cp_model.h"
#include "ortools/sat/cp_model.pb.h"
#include "ortools/sat/cp_model_solver.h"



namespace operations_research {
namespace sat {

void IntegerProgrammingExample() {
  

  

  const std::vector<std::vector<int>> costs{
      {90, 80, 75, 70},   {35, 85, 55, 65},   {125, 95, 90, 95},
      {45, 110, 95, 115}, {50, 100, 90, 100},
  };
  const int num_workers = static_cast<int>(costs.size());
  const int num_tasks = static_cast<int>(costs[0].size());
  


  

  

  CpModelBuilder cp_model;
  


  

  

  

  

  std::vector<std::vector<BoolVar>> x(num_workers,
                                      std::vector<BoolVar>(num_tasks));
  for (int i = 0; i < num_workers; ++i) {
    for (int j = 0; j < num_tasks; ++j) {
      x[i][j] = cp_model.NewBoolVar();
    }
  }
  


  

  

  

  for (int i = 0; i < num_workers; ++i) {
    cp_model.AddAtMostOne(x[i]);
  }
  

  for (int j = 0; j < num_tasks; ++j) {
    std::vector<BoolVar> tasks;
    for (int i = 0; i < num_workers; ++i) {
      tasks.push_back(x[i][j]);
    }
    cp_model.AddExactlyOne(tasks);
  }
  


  

  

  LinearExpr total_cost;
  for (int i = 0; i < num_workers; ++i) {
    for (int j = 0; j < num_tasks; ++j) {
      total_cost += x[i][j] * costs[i][j];
    }
  }
  cp_model.Minimize(total_cost);
  


  

  

  const CpSolverResponse response = Solve(cp_model.Build());
  


  

  

  if (response.status() == CpSolverStatus::INFEASIBLE) {
    LOG(FATAL) << "No solution found.";
  }

  LOG(INFO) << "Total cost: " << response.objective_value();
  LOG(INFO);
  for (int i = 0; i < num_workers; ++i) {
    for (int j = 0; j < num_tasks; ++j) {
      if (SolutionBooleanValue(response, x[i][j])) {
        LOG(INFO) << "Task " << i << " assigned to worker " << j
                  << ".  Cost: " << costs[i][j];
      }
    }
  }
  

}
}  

}  


int main(int argc, char** argv) {
  operations_research::sat::IntegerProgrammingExample();
  return EXIT_SUCCESS;
}

File: test/ok/assignment_sat.ts
import { operations_research as op } from '../../src'

function test()
{

    let costs = [
        [90, 80, 75, 70],
        [35, 85, 55, 65],
        [125, 95, 90, 95],
        [45, 110, 95, 115],
        [50, 100, 90, 100]
    ];
    let num_workers = costs.length;
    let num_tasks = costs[0].length;

    let cp_model = new op.sat.CpModelBuilder();

    let x = [];
    for (let i = 0; i < num_workers; ++i)
    {
        let row = [];
        for (let j = 0; j < num_tasks; ++j)
        {
            row.push(cp_model.NewBoolVar());
        }
        x.push(row);
    }

    for (let i = 0; i < num_workers; ++i)
    {
        cp_model.AddAtMostOne(x[i]);
    }


    for (let j = 0; j < num_tasks; ++j)
    {
        let tasks = [];
        for (let i = 0; i < num_workers; ++i)
        {
            tasks.push(x[i][j]);
        }
        cp_model.AddExactlyOne(tasks);
    }


    let total_cost = new op.sat.LinearExpr();
    for (let i = 0; i < num_workers; ++i)
    {
        for (let j = 0; j < num_tasks; ++j)
        {
            total_cost.operator_plus_eq(op.sat.operator_times(x[i][j], costs[i][j]));
        }
    }



    let response = op.sat.Solve(cp_model.Build());

    if (response.status() == op.sat.CpSolverStatus.INFEASIBLE)
    {
        console.log("No solution found.");
        return;
    }

    console.log("Total cost: " + response.objective_value());
    console.log();
    for (let i = 0; i < num_workers; ++i)
    {
        for (let j = 0; j < num_tasks; ++j)
        {
            if (op.sat.SolutionBooleanValue(response, x[i][j]))
            {
                console.log("Task " + i + " assigned to worker " + j +
                    ".  Cost: " + costs[i][j]);
            }
        }
    }
}

test();
